### PR TITLE
🐛 [RUMF-1060] fix failing worker detection in Firefox

### PR DESCRIPTION
**Pull request by BenoitZugmeyer**
## Motivation

In Firefox, when the worker fails to instanciate because of CSP rules, it doesn't throw an error like Chromium does, but rather emits a `error` event asynchronously.

## Changes

* Make sure we catch `error` events from the worker
* When creating the worker, check if the communication succeeds by sending an 'init' action and waiting for a 'initialized' response.

## Testing

CI, manual

---

I have gone over the [contributing](https://github.com/DataDog/browser-sdk/blob/main/CONTRIBUTING.md) documentation.

### CODE REVIEW

File: packages/rum/src/domain/segmentCollection/deflateWorkerSingleton.ts [hunk 1]
```diff
@@ -1,32 +1,97 @@
...
+      break
+  }
+}
+
+export function resetDeflateWorkerSingletonState() {
+  state = { status: DeflateWorkerSingletonStatus.Nil }
+}
+
+function startDeflateWorkerSingleton(createDeflateWorkerImpl: typeof createDeflateWorker) {
+  try {
```
Comment (amortemousque) [inline packages/rum/src/domain/segmentCollection/deflateWorkerSingleton.ts:55]: It could be interesting to add a comment on why they are a `try catch` and an `error event listener`. (Chromium/Firefox)

File: packages/rum/src/domain/segmentCollection/deflateWorkerSingleton.ts [hunk 2]
```diff
@@ -1,32 +1,97 @@
...
-      } else {
-        addErrorToMonitoringBatch(error)
-      }
-      return
+const enum DeflateWorkerSingletonStatus {
+  Nil,
+  Loading,
+  Error,
+  Ready,
+}
```
Comment (bcaudan) [inline packages/rum/src/domain/segmentCollection/deflateWorkerSingleton.ts:9]: It could be interesting to add some explanations about what we want to solve with these states

File: packages/rum/src/domain/segmentCollection/deflateWorkerSingleton.ts (hunk 1 above)
Comment (bcaudan) [reply packages/rum/src/domain/segmentCollection/deflateWorkerSingleton.ts:55]: it could be nice to check if there is some spec on the behavior and see if an issue should be opened somewhere

File: packages/rum/src/boot/recorderApi.spec.ts [hunk 3]
```diff
@@ -104,10 +107,22 @@ describe('makeRecorderApi', () => {
...
       expect(startRecordingSpy).not.toHaveBeenCalled()
     })
+
+    it('does not start recording multiple times if restarted before worker is ready', () => {
+      setupBuilder.build()
+      rumInit(DEFAULT_INIT_CONFIGURATION)
+      loadDeflateWorkerSingletonSpy.and.callFake(noop)
+      recorderApi.start()
+      recorderApi.stop()
+      loadDeflateWorkerSingletonSpy.calls.mostRecent().args[0]({} as DeflateWorker)
```
Comment (bcaudan) [inline packages/rum/src/boot/recorderApi.spec.ts:121]: the intent is a bit hard to follow, what about using a stub with an API explaining what is happening

File: packages/rum/src/boot/startRecording.spec.ts [hunk 4]
```diff
@@ -37,47 +37,50 @@ describe('startRecording', () => {
...
+        })
+        .beforeBuild(({ lifeCycle, applicationId, configuration, parentContexts, session }) => {
+          const recording = startRecording(
+            lifeCycle,
+            applicationId,
+            configuration,
+            session,
+            parentContexts,
+            deflateWorker!
+          )
```
Comment (bcaudan) [inline packages/rum/src/boot/startRecording.spec.ts:72]: what about using `startDeflateWorkerSingleton` to return a worker for the tests that need one?
IMO, no need to add more complexity to the setup

File: packages/rum/src/domain/segmentCollection/deflateWorkerSingleton.ts (hunk 1 above)
Comment (amortemousque) [reply packages/rum/src/domain/segmentCollection/deflateWorkerSingleton.ts]: I've opened a bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1736865

File: packages/rum/src/boot/recorderApi.spec.ts [hunk 5]
```diff
@@ -29,8 +45,9 @@ describe('makeRecorderApi', () => {
       startRecordingSpy = jasmine.createSpy('startRecording').and.callFake(() => ({
         stop: stopRecordingSpy,
       }))
-      getDeflateWorkerSingletonSpy = jasmine.createSpy('getDeflateWorkerSingleton').and.returnValue({})
-      recorderApi = makeRecorderApi(startRecordingSpy, getDeflateWorkerSingletonSpy)
+
+      startDeflateWorkerWith({} as DeflateWorker)
```
Comment (bcaudan) [inline packages/rum/src/boot/recorderApi.spec.ts:49]: what about extracting
```ts
const FAKE_WORKER = {} as DeflateWorker
```

File: packages/rum/src/boot/recorderApi.spec.ts [hunk 6]
```diff
@@ -19,7 +20,22 @@ describe('makeRecorderApi', () => {
...
+  let startDeflateWorkerSpy: jasmine.Spy<typeof startDeflateWorker>
+
+  function startDeflateWorkerWith(worker?: DeflateWorker) {
+    if (!startDeflateWorkerSpy) {
+      startDeflateWorkerSpy = jasmine.createSpy<typeof startDeflateWorker>('startDeflateWorker')
+    }
+    startDeflateWorkerSpy.and.callFake((callback) => callback(worker))
+  }
+
+  function callDeflateCallbackWith(worker?: DeflateWorker) {
```
Comment (bcaudan) [inline packages/rum/src/boot/recorderApi.spec.ts:32]: - do we need the worker parameter?
- what about renaming the function to something more domain related? maybe something like `triggerLastRegisteredInitialisationCallback()`

File: packages/rum/src/domain/segmentCollection/startDeflateWorker.spec.ts [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+describe('startDeflateWorker', () => {
+  let deflateWorker: MockWorker
+  let createDeflateWorkerSpy: jasmine.Spy<typeof createDeflateWorker>
+
+  beforeEach(() => {
+    resetDeflateWorkerState()
+    deflateWorker = new MockWorker()
+    createDeflateWorkerSpy = jasmine.createSpy().and.callFake(() => deflateWorker)
+    setDebugMode(true)
```
Comment (bcaudan) [inline packages/rum/src/domain/segmentCollection/startDeflateWorker.spec.ts:22]: debug mode does not seem to be needed

File: packages/rum/src/domain/segmentCollection/startDeflateWorker.spec.ts [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+import { MockWorker } from '../../../test/utils'
+import { createDeflateWorker } from './deflateWorker'
+import { startDeflateWorker, resetDeflateWorkerState } from './startDeflateWorker'
+
+describe('startDeflateWorker', () => {
+  let deflateWorker: MockWorker
+  let createDeflateWorkerSpy: jasmine.Spy<typeof createDeflateWorker>
+
+  beforeEach(() => {
+    resetDeflateWorkerState()
```
Comment (bcaudan) [inline packages/rum/src/domain/segmentCollection/startDeflateWorker.spec.ts:19]: the reset in afterEach should be enough, no?

File: packages/rum/src/domain/segmentCollection/startDeflateWorker.spec.ts [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    deflateWorker = new MockWorker()
+    createDeflateWorkerSpy = jasmine.createSpy().and.callFake(() => deflateWorker)
+    setDebugMode(true)
+  })
+
+  afterEach(() => {
+    resetDeflateWorkerState()
+  })
+
+  it('creates a deflate worker', () => {
```
Comment (bcaudan) [inline packages/rum/src/domain/segmentCollection/startDeflateWorker.spec.ts:29]:
```suggestion
  it('creates a deflate worker and call callback when initialized', () => {
```

File: packages/rum/src/domain/segmentCollection/startDeflateWorker.spec.ts [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+    const callbackSpy = jasmine.createSpy()
+    startDeflateWorker(callbackSpy, createDeflateWorkerSpy)
+    expect(createDeflateWorkerSpy).toHaveBeenCalledTimes(1)
+    deflateWorker.processAllMessages()
+    expect(callbackSpy).toHaveBeenCalledOnceWith(deflateWorker)
+  })
+
+  it('returns the previously created worker', () => {
```
Comment (bcaudan) [inline packages/rum/src/domain/segmentCollection/startDeflateWorker.spec.ts:37]:
```suggestion
  it('uses the previously created worker', () => {
```

File: packages/rum/src/domain/segmentCollection/startDeflateWorker.spec.ts [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+  })
+
+  describe('loading state', () => {
+    it('does not create multiple workers when called multiple times while the worker is loading', () => {
+      startDeflateWorker(noop, createDeflateWorkerSpy)
+      startDeflateWorker(noop, createDeflateWorkerSpy)
+      expect(createDeflateWorkerSpy).toHaveBeenCalledTimes(1)
+    })
+
+    it('calls all registered callbacks when the worker is loaded', () => {
```
Comment (bcaudan) [inline packages/rum/src/domain/segmentCollection/startDeflateWorker.spec.ts:55]:
```suggestion
    it('calls all registered callbacks when the worker is initialized', () => {
```

File: packages/rum/src/domain/segmentCollection/startDeflateWorker.spec.ts [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+    createDeflateWorkerSpy = jasmine.createSpy().and.callFake(() => deflateWorker)
+    setDebugMode(true)
+  })
+
+  afterEach(() => {
+    resetDeflateWorkerState()
+  })
+
+  it('creates a deflate worker', () => {
+    const callbackSpy = jasmine.createSpy()
```
Comment (bcaudan) [inline packages/rum/src/domain/segmentCollection/startDeflateWorker.spec.ts:30]: nit: since this spy is widely used, it could be initialized in the beforeEach

File: packages/rum/src/domain/segmentCollection/startDeflateWorker.spec.ts [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+import { startDeflateWorker, resetDeflateWorkerState } from './startDeflateWorker'
+
+describe('startDeflateWorker', () => {
+  let deflateWorker: MockWorker
+  let createDeflateWorkerSpy: jasmine.Spy<typeof createDeflateWorker>
+
+  beforeEach(() => {
+    resetDeflateWorkerState()
+    deflateWorker = new MockWorker()
+    createDeflateWorkerSpy = jasmine.createSpy().and.callFake(() => deflateWorker)
```
Comment (bcaudan) [inline packages/rum/src/domain/segmentCollection/startDeflateWorker.spec.ts:21]:
```suggestion
    createDeflateWorkerSpy = jasmine.createSpy('createDeflateWorkerSpy').and.callFake(() => deflateWorker)
```

nit: to ease debugging, it can be interesting to name the spies, especially when several are used in the same test

File: packages/rum/src/boot/startRecording.spec.ts [hunk 14]
```diff
@@ -61,23 +67,17 @@ describe('startRecording', () => {
...
-      waitAsyncCalls: waitRequestSendCalls,
-      expectNoExtraAsyncCall: expectNoExtraRequestSendCalls,
-    } = collectAsyncCalls(requestSendSpy))
   })
 
   afterEach(() => {
     sandbox.remove()
     setMaxSegmentSize()
-    setupBuilder.cleanup()
+    setupBuilder?.cleanup()
```
Comment (bcaudan) [inline packages/rum/src/boot/startRecording.spec.ts:80]:
```suggestion
    setupBuilder.cleanup()
```

File: packages/rum/src/domain/segmentCollection/startDeflateWorker.spec.ts [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+import { startDeflateWorker, resetDeflateWorkerState } from './startDeflateWorker'
+
+describe('startDeflateWorker', () => {
+  let deflateWorker: MockWorker
+  let createDeflateWorkerSpy: jasmine.Spy<typeof createDeflateWorker>
+  let callbackSpy: jasmine.Spy
+
+  beforeEach(() => {
+    deflateWorker = new MockWorker()
+    callbackSpy = jasmine.createSpy()
```
Comment (bcaudan) [inline packages/rum/src/domain/segmentCollection/startDeflateWorker.spec.ts:20]:
```suggestion
    callbackSpy = jasmine.createSpy('callbackSpy')
```
same as https://github.com/DataDog/browser-sdk/pull/1139#discussion_r733507038 then

File: packages/rum/src/domain/segmentCollection/startDeflateWorker.ts [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+
+/**
+ * Starts the deflate worker and handle messages and errors
+ *
+ * Browsers have discrepancies on how to handle worker errors:
+ * - Chromium throws an exception
+ * - Firefox fires an error event
+ *
+ * The spec is not very clear on how to handle worker CSP error.
+ * Here is the answer of Firefox about it: https://bugzilla.mozilla.org/show_bug.cgi?id=1736865#c2
```
Comment (bcaudan) [inline packages/rum/src/domain/segmentCollection/startDeflateWorker.ts:67]:
```suggestion
 * The spec allow browsers to handle worker errors differently:
 * - Chromium throws an exception
 * - Firefox fires an error event
 *
 * more details: https://bugzilla.mozilla.org/show_bug.cgi?id=1736865#c2
```

File: packages/rum/src/domain/segmentCollection/startDeflateWorker.ts [hunk 17]
```diff
@@ -0,0 +1,113 @@
+import { addErrorToMonitoringBatch, display, includes, monitor } from '@datadog/browser-core'
+import { createDeflateWorker, DeflateWorker } from './deflateWorker'
+
+/**
+ * For some browser like Firefox, "new Worker()" can fail and fire an 'error' event instead of an exception,
+ * making the creation asynchronous. To handle it, we need a round trip of initialization messages to be sure
+ * the creation is correct. These worker lifecycle states handle this case.
```
Comment (bcaudan) [inline packages/rum/src/domain/segmentCollection/startDeflateWorker.ts:7]:
```suggestion
In order to be sure that the worker is correctly working, we need a round trip of initialization messages, making the creation asynchronous.
These worker lifecycle states handle this case.
```
it seemed a bit redundant with the other comment

**Review by bcaudan — APPROVED** [review]
